app/utils/JourneyChecker.scala

```diff
@@ -32,23 +35,85 @@ trait JourneyChecker extends IncomeSourcesUtils {
   implicit val ec: ExecutionContext
 
   val sessionService: SessionService
-  def withIncomeSourcesFSWithSessionCheck(journeyType: JourneyType)(codeBlock: => Future[Result])(implicit user: MtdItUser[_], hc: HeaderCarrier): Future[Result] = {
-    withIncomeSourcesFS {
-      journeyChecker(journeyType).flatMap {
-        case true => user.userType match {
-          case Some(Agent) => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.showAgent(journeyType.businessType)))
-          case _ => Future.successful(Redirect(controllers.incomeSources.add.routes.YouCannotGoBackErrorController.show(journeyType.businessType)))
-        }
-        case false => codeBlock
+
+  private lazy val isAgent: MtdItUser[_] => Boolean = (user: MtdItUser[_]) => user.userType.contains(Agent)
+
+  private lazy val redirectUrl: (JourneyType, Option[String], Option[String], Option[String], Boolean) => MtdItUser[_] => Future[Result] =
+    (journeyType: JourneyType, reportingMethod: Option[String], taxYear: Option[String], incomeSourceId: Option[String], useDefault: Boolean) => user => {
```

Comment:
next parameters never actually used, if we can drop them:
`reportingMethod: Option[String], taxYear: Option[String], incomeSourceId: Option[String]`